snapcraft/internal/errors.py

```diff
@@ -23,22 +23,80 @@
 class SnapcraftError(Exception):
     """Base class for all snapcraft exceptions.
 
-    :cvar fmt: A format string that daughter classes override
-
+    :cvar brief: Concise, single-line description of the error.
+    :cvar details: Detailed information, if considered necessary for user.
+    :cvar resolution: Concise suggestion for user to resolve error.
+    :cvar docs_url: docs.snapcraft.io link, if applicable.
+    :cvar exit_code: Exit code to use if this exception causes Snapcraft to exit.
+    :cvar reportable: Prompt user to report this error to sentry.
+
+    Format string options may be used for templating keyworded exceptions.
+
+    :cvar fmt: (DEPRECATED) Legacy format string.
+    :cvar fmt_brief: Keyworded format string for error.
+    :cvar fmt_details: Keyworded format string for details.
+    :cvar fmt_resolution: Keyworded format string for resolution.
     """
 
-    fmt = "Daughter classes should redefine this"
+    fmt: str = ""
+    fmt_brief: str = ""
```

Comment:
if using AbstractBaseClass we can make the below properties (brief, ...) that are required to return formatted strings for this the base class to not be concerned about potentially unknown attributes.

Reply:
Something like?

```
class SnapcraftABCError(Exception, abc.ABC):
    """Base class for all snapcraft exceptions.

    :cvar brief: Concise, single-line description of the error.
    :cvar details: Detailed information, if considered necessary for user.
    :cvar resolution: Concise suggestion for user to resolve error.
    :cvar docs_url: docs.snapcraft.io link, if applicable.
    :cvar exit_code: Exit code to use if this exception causes Snapcraft to exit.
    :cvar reportable: Prompt user to report this error to sentry.
    """

    resolution: str = ""
    details: str = ""
    docs_url: str = ""
    exit_code: int = 2
    reportable: bool = False

    @property
    @abc.abstractmethod
    def brief(self) -> str:
        """Formatted error string using attributes as keywords."""
        pass

    @property
    def printable_message(self) -> str:
        """Formatted string suited to print to the user."""
        parts = [self.brief]

        if self.resolution:
            parts.extend(["", "Recommended resolution:", self.resolution])

        if self.details:
            parts.extend(["", "Detailed information:", self.details])

        if self.docs_url:
            parts.extend(["", "For more information, check out:", self.docs_url])

        return "\n".join(parts)

    def __str__(self):
        return self.printable_message

    def get_exit_code(cls) -> int:
        """Deprecated. Use exit_code class variable instead."""
        return cls.exit_code
```

Reply:
^ I'm unsure what else can be designated as abstract/required.  Perhaps resolution?

Reply:
Yes, but make those cvars actual properties too.

```python
class SanpcraftABCError(Exception, meta=abc.ABC):
    @property
    @abc.abstractmethod
    def resolution(self) -> str:
    """Return a string with the resolution... blah blah"""

same for details, docs_url, exit_code and reportable